### PR TITLE
Create recurring orders

### PR DESCRIPTION
**Pull request by dawehner**
Based upon https://github.com/bojanz/commerce_recurring/pull/18 right now

* [x] Allow subscription types to create the initial recurring orders and order items, referencing all the various bits
* [x] Wrote some basic test to ensure this is just happening on activation
* [x] Ensured that billing schedule types can determine the start and end date:

### CODE REVIEW

**Comment by bojanz** [comment]
Let's rebase this, I merged #18

**Comment by dawehner** [comment]
Rerolled and adapted the name

**Comment by dawehner** [comment]
Thank you @bojanz for the clarifying call. This PR is ready to review now. I'm updating the other one now, with hopefully a test attached.

File: src/Charge.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+   *
+   * @var \Drupal\commerce_price\Price
+   */
+  protected $amount;
+
+  /**
+   * Charge constructor.
+   *
+   * @param \Drupal\commerce_price\Price $amount
+   *   The amount of the charge.
```
Comment (bojanz) [inline src/Charge.php]: Why didn't we add the label, start/end dates right away?

File: src/Entity/Subscription.php [hunk 2]
```diff
@@ -227,6 +235,46 @@ public function setEndTime($timestamp) {
...
+   *   An array of default values.
+   */
+  public static function getCurrentUserId() {
+    return [\Drupal::currentUser()->id()];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getStores() {
```
Comment (bojanz) [inline src/Entity/Subscription.php]: getStores() looks like it was copied from somewhere by accident.

File: src/Entity/Subscription.php [hunk 3]
```diff
@@ -227,6 +235,46 @@ public function setEndTime($timestamp) {
...
+   * {@inheritdoc}
+   */
+  public function getOrderItemTypeId() {
+    return 'recurring';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getOrderItemTitle() {
```
Comment (bojanz) [inline src/Entity/Subscription.php]: Order item titles come from the charge, this looks unnecessary?

File: src/Plugin/Commerce/SubscriptionType/RepeatedOrder.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+use Drupal\commerce_recurring\Charge;
+use Drupal\commerce_recurring\Entity\SubscriptionInterface;
+
+/**
+ * @CommerceSubscriptionType(
+ *   id = "repeated_order",
+ *   label = @Translation("Repeated order"),
+ * )
+ */
+class RepeatedOrder extends SubscriptionTypeBase {
```
Comment (bojanz) [inline src/Plugin/Commerce/SubscriptionType/RepeatedOrder.php]: Why are we adding this subscription type? Why not use the existing product variation one?

Repeated orders might have additional complexity associated with them, so I kept them out of the initial scope.

File: src/Plugin/Commerce/SubscriptionType/SubscriptionTypeBase.php [hunk 5]
```diff
@@ -65,5 +71,69 @@ public function buildFieldDefinitions() {
...
+    ]);
+
+    foreach ($initial_charges as $charge) {
+      // Create the initial order item.
+      // @todo Take into account prepaid vs. postpaid
+      $order_item = $order_item_storage->createFromPurchasableEntity($subscription, [
+        'type' => 'recurring',
+        'billing_schedule' => $subscription->getBillingSchedule(),
+        'quantity' => 1,
+        'unit_price' => $charge->getAmount(),
```
Comment (bojanz) [inline src/Plugin/Commerce/SubscriptionType/SubscriptionTypeBase.php]: This is where we'd set the order item start/end dates as well, but I think we haven't created them yet.

File: src/Entity/Subscription.php (hunk 3 above)
Comment (dawehner) [reply src/Entity/Subscription.php]: See above.

File: src/Entity/Subscription.php (hunk 2 above)
Comment (dawehner) [reply src/Entity/Subscription.php]: These methods all come from ```\Drupal\commerce\PurchasableEntityInterface```

File: src/Plugin/Commerce/SubscriptionType/RepeatedOrder.php (hunk 4 above)
Comment (dawehner) [reply src/Plugin/Commerce/SubscriptionType/RepeatedOrder.php]: Fair

File: src/Entity/Subscription.php (hunk 2 above)
Comment (bojanz) [reply src/Entity/Subscription.php]: Ah, sorry about that.

File: src/Plugin/Commerce/SubscriptionType/SubscriptionTypeBase.php (hunk 5 above)
Comment (dawehner) [reply src/Plugin/Commerce/SubscriptionType/SubscriptionTypeBase.php]: Resolved now

**Comment by dawehner** [comment]
I just pushed addressing all the feedback from @bojanz which I forgot to push yesterday :(

**Pull request merged**